### PR TITLE
fixes #2135 - snapshots and versions installed by bnd workspace into a maven local repository are ignored by maven

### LINKED ISSUE #2135: snapshots and versions installed by bnd workspace into a maven local repository are ignored by maven
When a maven build searches for dependencies, if there is a remote repository which provides the version/snapshot, the local one is always ignored. The only workaround is to use offline mode.

### PR DESCRIPTION
**Pull request by rotty3000**

### CODE REVIEW

**Comment by bjhargrave** [comment]
Looks ok to me, but @pkriens needs to review.

**Comment by pkriens** [comment]
Don't have time to check it out but it looks harmless to me. So if it works ... I noticed there was no test case?

**Comment by rotty3000** [comment]
sure I can add a test!

All this change does is allow the metadata files to be written when a snapshot is built locally (i.e. not fetched from a repo).

**Comment by pkriens** [comment]
:-)

**Comment by rotty3000** [comment]
what this change boils down to is that even when building local snapshots, we need to create the metadata files. Caveat is that locally build snapshots have slightly different metadata.

File: biz.aQute.repository/src/aQute/maven/provider/Releaser.java [hunk 1]
```diff
@@ -71,24 +71,30 @@ protected void updateMetadata() throws Exception, InterruptedException {
 		int n = 0;
```
Comment (bjhargrave) [inline biz.aQute.repository/src/aQute/maven/provider/Releaser.java]: Seems like there is still an NPE risk here. You now alway call `updateMetadata` and it always calls `isUpdateProgramMetadata` whose implementation on `Releaser` assumes `repo` is non-null. I think you also need to fix `Releaser.isUpdateProgramMetadata` to guard against repo being null.

Comment (rotty3000) [reply biz.aQute.repository/src/aQute/maven/provider/Releaser.java]: no because the implementation overrides that method:
```
	protected boolean isUpdateProgramMetadata() {
		return true;
	}
```

Comment (rotty3000) [reply biz.aQute.repository/src/aQute/maven/provider/Releaser.java]: now i did consider that maybe the Snapshot releaser should just provide a completely different path for the non-repo case as opposed to having all these null checks when driving through the same logical path.

Comment (rotty3000) [reply biz.aQute.repository/src/aQute/maven/provider/Releaser.java]: ok... now I see what you mean...

Comment (rotty3000) [reply biz.aQute.repository/src/aQute/maven/provider/Releaser.java]: maybe there should be a new path for SnapshotReleaser with no repo?

Comment (bjhargrave) [reply biz.aQute.repository/src/aQute/maven/provider/Releaser.java]: That is possible. I thought of something like that. But I am not sure that is just too risky.

Comment (bjhargrave) [reply biz.aQute.repository/src/aQute/maven/provider/Releaser.java]: Also, I think you can release a non-snapshot version to a local only repository and so Releaser will be used with repo==null. So the fix cannot be only in SnapshotReleaser.

Comment (bjhargrave) [reply biz.aQute.repository/src/aQute/maven/provider/Releaser.java]: I think you only need up update the `Releaser.isUpdateProgramMetadata` method to handle repo==null. I dont think it is a big change.

**Comment by rotty3000** [comment]
the required change was much less intrusive in the end.

File: biz.aQute.repository/src/aQute/maven/provider/MetadataParser.java [hunk 2]
```diff
@@ -108,9 +108,12 @@ public Tag toTag() {
 			Tag versioning = new Tag(top, "versioning");
 
 			Tag snapshot = new Tag(versioning, "snapshot");
```
Comment (bjhargrave) [inline biz.aQute.repository/src/aQute/maven/provider/MetadataParser.java]: I think you need a test to not emit the snapshot tag if all of the fields are null. Now you can end up with an empty snapshot element.

Comment (rotty3000) [reply biz.aQute.repository/src/aQute/maven/provider/MetadataParser.java]: got it, the two sets are supposed to be kind of mutually exclusive.
